### ISSUE TITLE
Pixel to world coordinate transform consistency

### ISSUE THREAD

**Issue by orynaivashtenko** [lightkurve/lightkurve#1457]
#### Problem description

When using `lightkurve.targetpixelfile.KeplerTargetPixelFile.get_coordinates()`, a transformation is made therein:

```python
ra, dec = w.wcs_pix2world(X.ravel(), Y.ravel(), 1)
```
This transformation gives coordinates which are inconsistent with the Gaia coordinates. If one plots the coordinates of the pixels and the Gaia position of the source, one finds that the source is off the light in the pixels. (I used e.g Kepler KID 001577143 quarter 8).

I compared this to the result of the same plotting but in the pixel coordinates and not the sky coordinates, which is done in
`lightkurve.targetpixelfile.KeplerTargetPixelFile.interact_sky()`. It plots the pixels and the Gaia source position in the pixel coordinates, and everything is consistent, that is the source is consistent with the light of the pixels. In this function, the following transformation is used:

```python
radecs = np.vstack([result["RA_ICRS"], result["DE_ICRS"]]).T
coords = tpf.wcs.all_world2pix(radecs, 0)
```
It differs by the value `0` instead of `1` in the `origin` argument. 

#### Expected behavior

I think that value `0` should also be used in  `lightkurve.targetpixelfile.KeplerTargetPixelFile.get_coordinates()`. When I modified the code and tried it, I saw Gaia source in the correct place with the pixels in the ra-dec coordinates.

#### Environment

-  platform: Linux and OSX
-  lightkurve version: 2.4.2 
-  installation method: conda-forge

**Comment by vbkostov** [comment]
HI @orynaivashtenko , thank you for bringing this up. I am looking into the issue and would like to recreate your results. Can you please provide a full working example?

**Comment by orynaivashtenko** [comment]
> HI @orynaivashtenko , thank you for bringing this up. I am looking into the issue and would like to recreate your results. Can you please provide a full working example?


Here is the full working example. I calculate the centroid from the sky coordinates obtained through the lightkurve get_coordinates function (see the original issue description). I see that the centroid position is in offset with the Gaia position which I queried. When I modify the lightkurve function by selecting a different index convention for the origin, the centroid gets consistent with the Gaia position (see the plot)


```python
import os
import numpy as np
import warnings
import matplotlib.pyplot as plt
%matplotlib inline

from lightkurve import targetpixelfile as tpf


def my_centroid_series(target_pixel_fluxes, sky_coords=None, 
        aperture_masks=None, time_masks=None, dtype=np.float64):
    """
    This function calculates centroid through the moment method. 
    It uses sky coordinates if they are provided,
    otherwise returns the pixel coordinates. The aperture mask is 
    used if provided; if not, all the pixels are used.
    Beware that a too smell aperture mask does not make sense in 
    calculating the centroid, therefore it is
    better not to use the aperture photometry masks for centroids. 
    For example, the aperture photometry mask
    can be two pixels in y-direction, and then the centroid 
    in x-direction is not defined.
    
    :param target_pixel_fluxes: fluxes for all the quarters 
    for all the times for all the pixels
    :type target_pixel_fluxes: list of numpy.ndarrays of shape 
    (len_time, pixels_dim_1, pixels_dim_2) of float
    :param sky_coords: right ascensions and declinations for 
    all the quarters for all the times for all the pixels.
        If None, pixel indices (not true detector coordinates!) 
        will be used
    :type sky_coords: None or list of numpy.ndarrays of shape 
    (2, len_time, pixels_dim_1, pixels_dim_2)
    :param aperture_masks: mask of pixels that were used in 
    the optimal aperture to calculate the SAP flux.
    :type aperture_masks: list of numpy.ndarrays of shape 
    (pixels_dim_1, pixels_dim_2)
    :param time_masks: mask of times that are bad, including 
    the Kepler quality mask
    :type time_masks: list of numpy.ndarrays of shape (len_time)
    :return:
    :rtype:
    """
    target_pixel_fluxes = [np.array(e, dtype=dtype) for e in target_pixel_fluxes]
    centroids_x = []  # ra if sky_coords is specified
    centroids_y = []  # dec if sky_coords is specified
    for q in range(len(target_pixel_fluxes)):
        target_pixels_flux = target_pixel_fluxes[q]
        nan_aperture_masks = ~((np.isnan(target_pixels_flux)).all(axis=0))
        if aperture_masks is None:
            aperture_mask = nan_aperture_masks
        else:
            aperture_mask = (aperture_masks[q] * nan_aperture_masks).astype(bool)

        nan_mask = ~((np.isnan(target_pixels_flux)).all(axis=(1, 2)))
        if time_masks is None:
            time_mask = nan_mask
        else:
            time_mask = (time_masks[q] * nan_mask).astype(bool)

        if sky_coords is None:
            y_coor, x_coor = np.indices(np.shape(target_pixels_flux)[1:3])
            assert ((y_coor[aperture_mask] / np.mean(y_coor[aperture_mask])) != 1.
                    ).any(), 'All the pixels are in the same row! Centroid is constant'
            assert ((x_coor[aperture_mask] / np.mean(x_coor[aperture_mask])) != 1.
                    ).any(), 'All the pixels are in the same column! Centroid is constant'
            x_coor = x_coor * np.ones(np.shape(target_pixels_flux), dtype=dtype)
            y_coor = y_coor * np.ones(np.shape(target_pixels_flux), dtype=dtype)
        else:
            y_coor, x_coor = sky_coords[q][1], sky_coords[q][0]

        sap_flux = np.zeros(len(target_pixels_flux), dtype=dtype) * np.nan
        centr_x = np.zeros(len(target_pixels_flux), dtype=dtype) * np.nan
        centr_y = np.zeros(len(target_pixels_flux), dtype=dtype) * np.nan
        sap_flux[time_mask] = np.nansum(target_pixels_flux[time_mask][:, aperture_mask], 
                                        axis=1)
        centr_x[time_mask] = np.nansum((x_coor[time_mask] * target_pixels_flux[time_mask]
                                        )[:, aperture_mask], axis=1) / sap_flux[time_mask]
        centr_y[time_mask] = np.nansum((y_coor[time_mask] * target_pixels_flux[time_mask]
                                        )[:, aperture_mask], axis=1) / sap_flux[time_mask]
        centroids_x.append(centr_x)
        centroids_y.append(centr_y)
    return centroids_x, centroids_y


def lightkurve_get_coordinates_modified(self, cadence="all", wcs_origin=1):
    """
    copy of the lightkurve function lightkurve.KeplerTargetPixelFile.get_coordinates 
    but with one modification.
    MODIFICATION: added the `wcs_origin` option.
    `wcs_origin`=1 is the default Lightkurve behavior.
    I added option of 0, which shifts the result by one pixel, and I find that 
    it is correct because the result is consistent with the astrometric 
    positions of the sources.
    
    Returns two 3D arrays of RA and Dec values in decimal degrees.

    If cadence number is given, returns 2D arrays for that cadence. If
    cadence is 'all' returns one RA, Dec value for each pixel in every cadence.
    Uses the WCS solution and the POS_CORR data from TPF header.

    Parameters
    ----------
    cadence : 'all' or int
        Which cadences to return the RA Dec coordinates for.

    Returns
    -------
    ra : numpy array, same shape as tpf.flux[cadence]
        Array containing RA values for every pixel, for every cadence.
    dec : numpy array, same shape as tpf.flux[cadence]
        Array containing Dec values for every pixel, for every cadence.
    """
    w = self.wcs
    X, Y = np.meshgrid(np.arange(self.shape[2]), np.arange(self.shape[1]))
    pos_corr1_pix = np.copy(self.hdu[1].data["POS_CORR1"])
    pos_corr2_pix = np.copy(self.hdu[1].data["POS_CORR2"])

    # We zero POS_CORR* when the values are NaN or make no sense (>50px)
    with warnings.catch_warnings():  # Comparing NaNs to numbers is OK here
        warnings.simplefilter("ignore", RuntimeWarning)
        bad = np.any(
            [
                ~np.isfinite(pos_corr1_pix),
                ~np.isfinite(pos_corr2_pix),
                np.abs(pos_corr1_pix - np.nanmedian(pos_corr1_pix)) > 50,
                np.abs(pos_corr2_pix - np.nanmedian(pos_corr2_pix)) > 50,
            ],
            axis=0,
        )
    pos_corr1_pix[bad], pos_corr2_pix[bad] = 0, 0


    X = np.atleast_3d(X).transpose([2, 0, 1]) + np.atleast_3d(
        pos_corr1_pix
    ).transpose([1, 2, 0])
    Y = np.atleast_3d(Y).transpose([2, 0, 1]) + np.atleast_3d(
        pos_corr2_pix
    ).transpose([1, 2, 0])        

    # Pass through WCS
    ra, dec = w.wcs_pix2world(X.ravel(), Y.ravel(), wcs_origin)
    ra = ra.reshape((pos_corr1_pix.shape[0], self.shape[1], self.shape[2]))
    dec = dec.reshape((pos_corr2_pix.shape[0], self.shape[1], self.shape[2]))
    ra, dec = ra[self.quality_mask], dec[self.quality_mask]

    if cadence != "all":
        return ra[cadence], dec[cadence]
    return ra, dec



# -----------------------------------------------------------------------------
# EXAMPLE:
# KIC 001577143, quarter 8
tid = '001577143'
Target_pixels_location = '/home/orynai/Kepler_Notebooks/TargetPixelsTemp'
tp_data_location = os.path.join(Target_pixels_location, tid[:4], tid)
tpf1 = tpf.KeplerTargetPixelFile(os.path.join(tp_data_location, os.listdir(tp_data_location)[8]))
# -----------------------------------------------------------------------------
plt.figure()
for wcs_origin in [0, 1]:
    ra, dec = lightkurve_get_coordinates_modified(tpf1, cadence="all", wcs_origin=wcs_origin)
    centroids_x, centroids_y = my_centroid_series(target_pixel_fluxes=[tpf1.flux.value], sky_coords=[[ra, dec]], aperture_masks=[tpf1.pipeline_mask], time_masks=None)
    plt.plot(centroids_x[0], centroids_y[0], 'o', markersize=2, label='wcs_origin ' + str(wcs_origin))

# I Querried this object in Gaia DR3:
ra_gaia = 292.1760759233577
dec_gaia = 37.119353567976475
plt.plot([ra_gaia], [dec_gaia], 'o', markersize=5, color='k', label='Gaia')

plt.legend(loc='best')
plt.xlabel('centroid ra, deg')
plt.ylabel('centroid dec, deg')

```

The output:

![image](https://github.com/user-attachments/assets/0c31978a-f2a0-48de-aed5-b61bf498c63c)



![image](https://github.com/user-attachments/assets/3f974ba9-f33c-482b-9720-5669b2316e0e)

**Comment by vbkostov** [comment]
Thank you for again for raising this issue, and for providing a full working example. It looks like this is indeed an issue that goes unnoticed unless one uses TargetPixelFile.get_coordinates or TargetPixelFile.cutout. We will patch it asap

**Comment by Nschanche** [comment]
I've merged this fix in PR#1465, so I'm closing this issue.
